tests/tracing/test_instrument_tracing.py

```diff
@@ -0,0 +1,153 @@
+    span = out_spans[0]
+    assert span.name == "func_with_carrier"
+    assert span.context.trace_id == int(root_trace_id, 16)
+    assert span.parent.span_id == int(root_span_id, 16)
+
+
+def test_instrument_func_multiple_threads(monkeypatch, fake_span_exporter):
+    root_trace_id = "cefb2b8db35d5f3c0dfdf79d5aab1451"
+    root_span_id = "1f2bb7927f140744"
+
+    monkeypatch.setenv("OTEL_TRACING", "true")
+    monkeypatch.setenv("OTEL_SERVICE_NAME", "local-test1")
+    monkeypatch.setenv("traceparent", f"00-{root_trace_id}-{root_span_id}-01")
+
+    tw = get_trace_wrapper()
+
+    @tw.instrument_func(span_name="sub_thread_span")
+    def sub_thread():
+        return 1
+
+    @tw.instrument_func(span_name="main_thread_span", args_to_attr=True)
+    def main_thread(param1, param2):
+        with ThreadPoolExecutor(max_workers=1) as executor:
+            future_res = [executor.submit(sub_thread) for i in range(1, 2)]
+            as_completed(future_res)
+
+    main_thread("p1", param2="p2")
+    tw.processor.force_flush()
+
+    main_thread_span = None
+    sub_thread_span = None
+    for span in tw.processor.span_exporter.get_spans():
+        if span.name == "main_thread_span":
+            main_thread_span = span
+        if span.name == "sub_thread_span":
+            sub_thread_span = span
+
+    assert main_thread_span
+    assert sub_thread_span
+
+    # Both are in the same trace
+    assert (
+        sub_thread_span.context.trace_id
+        == main_thread_span.context.trace_id
+        == int(root_trace_id, 16)
+    )
+
+    # Verify span parent and child relation
+    assert sub_thread_span.parent.span_id == main_thread_span.context.span_id
+    assert main_thread_span.parent.span_id == int(root_span_id, 16)
+
+    # Assert function parameters in span attributes
+    assert (
+        "args" in main_thread_span.attributes
+        and main_thread_span.attributes["args"] == "p1"
+    )
+    assert (
+        "kwargs" in main_thread_span.attributes
+        and main_thread_span.attributes["kwargs"] == "param2=p2"
+    )
+
+
+def test_instrument_func_exception(monkeypatch, fake_span_exporter):
+    monkeypatch.setenv("OTEL_TRACING", "true")
+    monkeypatch.setenv("OTEL_SERVICE_NAME", "local-test3")
+
+    tw = get_trace_wrapper()
+
+    @tw.instrument_func(span_name="func_with_exception")
+    def func_with_exception():
+        raise Exception("failed with exception")
+
+    assert tw
+    with pytest.raises(Exception):
+        func_with_exception()
+
+    tw.processor.force_flush()
+    out_spans = tw.processor.span_exporter.get_spans()
+
+    assert len(out_spans) == 1
+    span = out_spans[0]
+    assert span.name == "func_with_exception"
+    assert span.status.status_code == StatusCode.ERROR
+    assert len(span.events) >= 1
+    assert span.events[0].attributes["exception.message"] == "failed with exception"
+
+
+def test_instrument_func_disabled(monkeypatch):
+    monkeypatch.setenv("OTEL_TRACING", "false")
```

Comment:
I think this test is a bit faulty.  It forces OTEL_TRACING to false, but if a previous test has run and initialized the singleton and not wiped it out at the end of the test, then `get_trace_wrapper()` below is still going to return the object created during other tests.

If this test doesn't crash with tw being None, maybe that is the reason?  You might be able to reproduce it if you use pytest to run just this testfunction alone, rather than running the entire file.

Reply:
You are right.
Is there any way to wipe singleton and recreate a new one in tests?

Reply:
Updated TracingWrapper to able to be wiped out so that disabling OTEL tracing can be tested.